src/sentry/api/endpoints/grouping_level_details.py

```diff
@@ -0,0 +1,264 @@
+        been started, it will take a couple seconds to observe the effect. No
+        read-your-writes here.
+        """
+
+        check_feature(group.project.organization, request)
+
+        parsed_id = int(id)
+        levels_overview = get_levels_overview(group)
+
+        if parsed_id >= levels_overview.num_levels or parsed_id < 0:
+            raise InvalidLevel()
+
+        if parsed_id < levels_overview.current_level:
+            _decrease_level(group, parsed_id, levels_overview, request)
+
+        if parsed_id > levels_overview.current_level:
+            _increase_level(group, parsed_id, levels_overview, request)
+
+        return self.respond(status=200)
+
+
+def _increase_level(group: Group, id: int, levels_overview: LevelsOverview, request):
+    for other_group_id, state in GroupHash.objects.filter(
+        project_id=group.project_id, hash__in=levels_overview.parent_hashes
+    ).values_list("group_id", "state"):
+        if other_group_id is not None and state != GroupHash.State.SPLIT:
+            logger.error("grouping_levels.stacked_groups", extra={"group_id": group.id})
+
+    # Mark one hierarchical hash as SPLIT. Note this also prevents it from
+    # being deleted in group deletion.
+    #
+    # We're upserting the hash here to make sure it exists. We have
+    # observed that the materialized hash in postgres is mysteriously lost,
+    # presumably because of secondary grouping or merge/unmerge.
+    grouphash, _created = GroupHash.objects.update_or_create(
+        project_id=group.project_id,
+        hash=levels_overview.current_hash,
+        defaults={"state": GroupHash.State.SPLIT, "group_id": None},
+    )
+
+    replacement = HierarchicalUnmergeReplacement(
+        primary_hash=levels_overview.only_primary_hash,
+        filter_hierarchical_hash=levels_overview.current_hash,
+        filter_level=levels_overview.current_level,
+        new_level=id,
+    )
+
+    unmerge.delay(
+        project_id=group.project_id,
+        source_id=group.id,
+        destination_id=None,
+        fingerprints=None,
+        actor_id=request.user.id if request.user else None,
+        replacement=replacement,
+    )
+
+
+def _decrease_level(group: Group, id: int, levels_overview: LevelsOverview, request):
+    assert levels_overview.parent_hashes
+    assert id < len(levels_overview.parent_hashes)
+
+    new_materialized_hash = levels_overview.parent_hashes[id]
+
+    timerange = _bogus_timestamp_conditions()
+
+    # When decreasing the level, events from this + other groups get merged
+    # into one new parent group. This is why we cannot restrict the query by
+    # timestamp here. The query logic is analoguous to
+    # GroupingLevelNewIssuesEndpoint.
+
+    query = (
+        Query("events", Entity("events"))
+        .set_select(
+            [
+                Function(
+                    "arrayDistinct", [Function("groupArray", [Column("group_id")])], "group_ids"
+                ),
+                Function(
```

Comment:
@tkaemming this query is nuts but I believe it may be the fastest way to compute two unrelated queries at the same time: `SELECT distinct group_id ...` and `SELECT distinct arrayJoin(hierarchical_hashes) ...`. I suppose it would be better to send off two queries in a threadpool?